Project/Assets/ML-Agents/Examples/SharedAssets/Scripts/TargetController.cs

```diff
@@ -0,0 +1,143 @@
+﻿using UnityEngine;
+using Random = UnityEngine.Random;
+using Unity.MLAgents;
+using UnityEngine.Events;
+
+namespace Unity.MLAgentsExamples
+{
+    /// <summary>
+    /// Utility class to allow target placement and collision detection with an agent
+    /// Add this script to the target you want the agent to touch.
+    /// Callbacks will be triggered any time the target is touched with a collider tagged as 'tagToDetect'
+    /// </summary>
+    public class TargetController : MonoBehaviour
+    {
+        
+        [Header("Collider Tag To Detect")]
+        public string tagToDetect = "agent"; //collider tag to detect 
+        
+        [Header("Target Placement")]
+        public float spawnRadius; //The radius in which a target can be randomly spawned.
+        public bool respawnIfTouched; //Should the target respawn to a different position when touched
+        
+        [Header("Target Fell Protection")]
+        public bool respawnIfFallsOffPlatform = true; //If the target falls off the platform, reset the position.
+        public float fallDistance = 5; //distance below the starting height that will trigger a respawn 
+
+
+        private Vector3 m_startingPos; //the starting position of the target
+        private Agent m_agentTouching; //the agent currently touching the target
+
+        [System.Serializable]
+        public class TriggerEvent : UnityEvent<Collider>
+        {
+        }
+
+        [Header("Trigger Callbacks")] public bool triggerIsTouching;
+        public TriggerEvent onTriggerEnterEvent = new TriggerEvent();
+        public TriggerEvent onTriggerStayEvent = new TriggerEvent();
+        public TriggerEvent onTriggerExitEvent = new TriggerEvent();
+
+        [System.Serializable]
+        public class CollisionEvent : UnityEvent<Collision>
+        {
+        }
+
+        [Header("Collision Callbacks")] public bool colliderIsTouching;
+        public CollisionEvent onCollisionEnterEvent = new CollisionEvent();
```

Comment:
This is the only event we use and I think we should register the Callback to TouchedTarget via code rather than via UI

Reply:
I think the ultimate purpose of this script is to handle as many of our current "Target" situations as possible instead of ad-hoc implementations. if we leave the other callbacks we can use this script for other things like Reacher which works with a OnTriggerStay() right now. This script could replace the ReacherGoal.cs.

Reply:
I like this script, the thing I would like is 
```
        [Header("Collision Callbacks")]
        public CollisionEvent onCollisionEnterEvent = new CollisionEvent();
        public CollisionEvent onCollisionStayEvent = new CollisionEvent();
        public CollisionEvent onCollisionExitEvent = new CollisionEvent();
```
and 
```
        [Header("Trigger Callbacks")]
        public TriggerEvent onTriggerEnterEvent = new TriggerEvent();
        public TriggerEvent onTriggerStayEvent = new TriggerEvent();
        public TriggerEvent onTriggerExitEvent = new TriggerEvent();
```
Keep only one (onCollisionEnterEvent), make it private / hide in inspector. Add a public method to subscribe a callback to this event. Have the Agent subscribe to the callback via code (since it already has a reference to the "TargetController"). This simplifies the inspector UI and makes it so that a user can understand what happens by looking at the code.

Reply:
I'm not necessarily against doing this, but it goes against my original design goal which is to reduce the amount of code that needs to be in the Agent script. We'd need to Subscribe/Unsubscribe via code in each agent script. Exposing these events in the editor gets rid of all of that and gives us a general use component

Reply:
All right then